include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+        [[nodiscard]] IN6_ADDR* in6_addr() WI_NOEXCEPT;
+
+        [[nodiscard]] const SOCKADDR* sockaddr() const WI_NOEXCEPT;
+        [[nodiscard]] const SOCKADDR_IN* sockaddr_in() const WI_NOEXCEPT;
+        [[nodiscard]] const SOCKADDR_IN6* sockaddr_in6() const WI_NOEXCEPT;
+        [[nodiscard]] const SOCKADDR_INET* sockaddr_inet() const WI_NOEXCEPT;
+        [[nodiscard]] const IN_ADDR* in_addr() const WI_NOEXCEPT;
+        [[nodiscard]] const IN6_ADDR* in6_addr() const WI_NOEXCEPT;
+
+        [[nodiscard]] SOCKADDR_STORAGE sockaddr_storage() const WI_NOEXCEPT;
+
+        [[nodiscard]] constexpr int length() const
+        {
+            return sizeof(m_sockaddr);
+        }
+
+    private:
+        SOCKADDR_INET m_sockaddr{};
+    };
+
+    // When using dual-mode sockets, one might need to connect to a target IPv4 address.
+    // Since dual-mode socket types are IPv6, one must map that IPv4 address to its 'mapped' IPv6 address
+    inline ::wil::network::socket_address map_dual_mode_4to6(const ::wil::network::socket_address& ipv4_address) WI_NOEXCEPT
+    {
+        constexpr IN6_ADDR ipv4MappedPrefix{{IN6ADDR_V4MAPPEDPREFIX_INIT}};
+
+        ::wil::network::socket_address return_ipv6_address{&ipv4MappedPrefix, ipv4_address.port()};
+
+        auto* const return_in6_addr{return_ipv6_address.in6_addr()};
+        const auto* const ipv4_inaddr{ipv4_address.in_addr()};
+        return_in6_addr->u.Byte[12] = ipv4_inaddr->S_un.S_un_b.s_b1;
+        return_in6_addr->u.Byte[13] = ipv4_inaddr->S_un.S_un_b.s_b2;
+        return_in6_addr->u.Byte[14] = ipv4_inaddr->S_un.S_un_b.s_b3;
+        return_in6_addr->u.Byte[15] = ipv4_inaddr->S_un.S_un_b.s_b4;
+
+        return return_ipv6_address;
+    }
+
+    // non-member swap
+    inline void swap(::wil::network::socket_address& lhs, ::wil::network::socket_address& rhs) WI_NOEXCEPT
+    {
+        lhs.swap(rhs);
+    }
+
+    // class addr_info encapsulates the ADDRINFO-related structures returned from the socket functions
+    // getaddrinfo, GetAddrInfoW, GetAddrInfoWEx
+    // iterator semantics are supported to safely access these addresses
+    // template T supports pointers to the 3 address structures: ADDRINFOA*, ADDRINFOW*, ADDRINFOEXW*
+    template <typename T>
+    class addr_info_iterator_t
+    {
+    public:
+        // defining iterator_traits allows STL <algorithm> functions to be used with this iterator class.
+        // Notice this is a forward_iterator
+        // - does not support random-access (e.g. vector::iterator)
+        // - does not support bidirectional access (e.g. list::iterator)
+#if WIL_USE_STL || defined(WIL_DOXYGEN)
+        using iterator_category = ::std::forward_iterator_tag;
+#endif
+        using value_type = ::wil::network::socket_address;
+        using difference_type = size_t;
+        using distance_type = size_t;
+        using pointer = ::wil::network::socket_address*;
+        using reference = ::wil::network::socket_address&;

```

Comment:
```suggestion
        using pointer = const ::wil::network::socket_address*;
        using reference = const ::wil::network::socket_address&;
```
Your operators return `const` pointers/references. These types should reflect that